apps/ejabberd/src/mongoose_api_backend.erl

```diff
@@ -0,0 +1,234 @@
+
+to_json(Req, State) ->
+    handle_request(mongoose_api_json, Req, State).
+
+%%--------------------------------------------------------------------
+from_json(Req, State) ->
+    handle_body(mongoose_api_json, Req, State).
+
+%%--------------------------------------------------------------------
+%% internal funs
+%%--------------------------------------------------------------------
+
+handle_request(Serializer, Req, #backend_state{bindings=Bindings, command_category = Name}=State) ->
+    %% should we introduce filtering not only by category?
+    Commands = ?COMMANDS_ENGINE:list(admin, Name),
+    [Command] = [C || C <- Commands, ?COMMANDS_ENGINE:action(C) =:= read],
+    Result = execute_command(extract_bindings(Bindings), Command),
+    handle_result(Result, Serializer, Req, State).
+
+handle_body(Deserializer, Req, State) ->
+    {Method, Req1} = cowboy_req:method(Req),
+    {ok, Body, Req2} = cowboy_req:body(Req1),
+    case Deserializer:deserialize(Body) of
+        {ok, Data} ->
+            handle_body(Method, Data, Req2, State);
+        {error, _Reason} ->
+            error_response(bad_request, Req2, State)
+    end.
+
+handle_body(Method, Data, Req, #backend_state{command_category = Category}=State) ->
+    Action = method_to_action(Method),
+    Commands = ?COMMANDS_ENGINE:list(admin, Category),
+    [Command] = [C || C <- Commands, ?COMMANDS_ENGINE:action(C) =:= Action],
+    case check_and_extract_args(?COMMANDS_ENGINE:args(Command), Data) of
+        {ok, Args} ->
+            Result = execute_command(Args, Command),
+            handle_result(Result, Req, State);
+        {error, Type} ->
+            error_response(Type, Req, State)
+    end.
+
+handle_result({ok, Result}, Serializer, Req, State) ->
+    serialize(Result, Serializer, Req, State);
+handle_result(Other, _Serializer, Req, State) ->
+    handle_result(Other, Req, State).
+
+handle_result({ok, _Res}, Req, State) ->
+    %% TODO When POST add resource created to headers
+    %% TODO return appropriate status code
+    {ok, Req2} = cowboy_req:reply(201, [{<<"location">>, <<"http://localhost:5288/api/user">>}], Req),
+    {halt, Req2, State};
+handle_result({error, Error}, Req, State) ->
+    error_response(Error, Req, State);
+handle_result(no_call, Req, State) ->
+    error_response(not_implemented, Req, State).
+
+serialize(Data, Serializer, Req, State) ->
+    {Serializer:serialize(Data), Req, State}.
+
+register_handler(Base, Command) ->
+    {[Base, create_url_path(Command)],
+      ?MODULE, [{command_category, ?COMMANDS_ENGINE:category(Command)}]}.
+
+get_allowed_methods() ->
+    Commands = ?COMMANDS_ENGINE:list(admin),
+    [action_to_method(?COMMANDS_ENGINE:action(Command)) || {_Name, Command} <- Commands].
+
+execute_command(Args, Command) ->
+    ?COMMANDS_ENGINE:execute(admin, ?COMMANDS_ENGINE:name(Command), Args).
+
+create_url_path(Command) ->
+    "/" ++ maybe_add_bindings(category_to_resource(?COMMANDS_ENGINE:category(Command)), Command).
```

Comment:
maybe_add_bindings returns category part plus some bindings if present, and only the latter is the "maybe" part. Rephrased like: "/" ++ category_to_resoure ++ maybe_bindings would make it much more readable.